pupepat/test/pupil-synth/src/source.py

```diff
@@ -0,0 +1,157 @@
+import abc
+import numpy as np
+
+
+class Source(abc.ABC):
+    def __init__(self, x=None, y=None):
+        """
+        A Source need not know it's position on a plate until it's been added.
+        :param x:
+        :param y:
+        """
+        self.x, self.y = x, y
+        self.set_position(x, y)
+        self.half_width = 0  # decal.shape[0] = 2*half_width + 1
+        super().__init__()
+
+    def set_position(self, x, y):
+        self.x = x
+        self.y = y
+
+    @property
+    def position(self):
+        return self.x, self.y
+
+    @property
+    @abc.abstractmethod
+    def decal(self):
+        pass
+
+
+class PointSource(Source):
+    def __init__(self, radius, counts, x=None, y=None):
+        self.radius = radius
+        self.counts = counts
+        super().__init__(x, y)
+
+        self.half_width = radius  # same as radius (for PointSource)
+
+    def __str__(self):
+        return 'PointSource(radius={radius:0.2f}, counts={counts:0g}, ' \
+               'x={x:}, y={y:})'.format(radius=self.radius, counts=self.counts,
+                                        x=self.x, y=self.y)
+
+    @property
+    def decal(self):
+        shape = (2*self.radius + 1, 2*self.radius + 1)
+        _decal = np.zeros(shape)
+
+        ci, cj = self.half_width, self.half_width  # center of decal
+        ix, jx = np.meshgrid(np.arange(shape[0]), np.arange(shape[1]))
+
+        # distance points in the index arrays (ix, jx) to the decal center (ci, cj)
+        dist = np.sqrt((ix-ci)**2 + (jx-cj)**2)
+        _decal[np.where(dist < self.radius)] = self.counts  # fill the circle with counts
+        return _decal
+
+
+class CircularPupil(Source):
+    def __init__(self, inner_radius, outer_radius, counts,
+                 center_vector=(0, 0), x=None, y=None):
+        self.inner_radius = inner_radius
+        self.outer_radius = outer_radius
+        self.counts = counts
+        self.center_vector = center_vector
+        super().__init__(x, y)
```

Comment:
Since x and y are kwargs instead of args in the superclass, it is generally better to x=x, y=y.